<a href="https://colab.research.google.com/github/nfathimmain-sketch/python-assessment/blob/main/Data_acquisition_casestudy_fathimaN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


TASK 1

Read the Population dataset from the CSV file.

Hint:

● Use the pandas library.

● Display the first five rows.

● Check the shape and data types.

In [7]:
csv_data = pd.read_csv("/content/drive/MyDrive/DSA/Data_acquisition_casestudy_fathimaN/population.csv")
csv_data.head()

,country,population
0,United States,331002651
1,China,1411778724
2,India,1380004385
3,Japan,125836021
4,Germany,83240525


In [8]:
#Check the shape and data types
print(csv_data.shape)
print(csv_data.dtypes)

(25, 2)
country       object
population     int64
dtype: object


Task 2

Read the GDP dataset from the Excel file.

Hint:

● Use the appropriate pandas function.

● Display column names and data types.

In [10]:
excel_data = pd.read_excel("/content/drive/MyDrive/DSA/Data_acquisition_casestudy_fathimaN/gdp.xlsx")
#Check the column names and data types
print(excel_data.columns.tolist())
print(excel_data.dtypes)

['Country', 'GDP']
Country    object
GDP         int64
dtype: object


Task 3

Read the Internet Users dataset from the JSON file.

Hint:

● Use either the json module or pandas.

● Convert the data into a DataFrame.

In [11]:
json_data = pd.read_json("/content/drive/MyDrive/DSA/Data_acquisition_casestudy_fathimaN/internet_users.json")
pd.DataFrame(json_data)

,country,internet_users
0,United States,312000000
1,China,1080000000
2,India,850000000
3,Japan,118000000
4,Germany,79000000
5,United Kingdom,66000000
6,France,62000000
7,Italy,56000000
8,Canada,36000000
9,Australia,24000000


Task 4

Read the Literacy Rate dataset from the XML file.

Hint:

● Use xml.etree.ElementTree or pandas.read_xml().

● Extract all required fields into a DataFrame.




In [12]:
%pip install lxml
xml_data = pd.read_xml("/content/drive/MyDrive/DSA/Data_acquisition_casestudy_fathimaN/literacy_rate.xml")
pd.DataFrame(xml_data)

,name,literacy_rate
0,United States,99.0
1,China,97.2
2,India,77.7
3,Japan,99.0
4,Germany,99.0
5,United Kingdom,99.0
6,France,99.0
7,Italy,99.0
8,Canada,99.0
9,Australia,99.0


Task 5

Explore each dataset. Find the following:

● Number of rows

● Number of columns

● Column names

● Data types

● Missing values

● Duplicate records

In [13]:
#Population

print("No. of rows: ",len(csv_data))
print("No. of Columns: ",len(csv_data.columns))
print("Column Names: ",csv_data.columns.tolist())
print("Data types: ",csv_data.dtypes)
print("Missing values: ",csv_data.isnull().sum().tolist())
print("Duplicate values: ",csv_data.duplicated().sum())

No. of rows:  25
No. of Columns:  2
Column Names:  ['country', 'population']
Data types:  country       object
population     int64
dtype: object
Missing values:  [0, 0]
Duplicate values:  0


In [14]:
#GDP

print("No. of rows: ",len(excel_data))
print("No. of Columns: ",len(excel_data.columns))
print("Column Names: ",excel_data.columns.tolist())
print("Data types: ",excel_data.dtypes)
print("Missing values: ",excel_data.isnull().sum().tolist())
print("Duplicate values: ",excel_data.duplicated().sum())

No. of rows:  25
No. of Columns:  2
Column Names:  ['Country', 'GDP']
Data types:  Country    object
GDP         int64
dtype: object
Missing values:  [0, 0]
Duplicate values:  0


In [15]:
#Internet users

print("No. of rows: ",len(json_data))
print("No. of Columns: ",len(json_data.columns))
print("Column Names: ",json_data.columns.tolist())
print("Data types: ",json_data.dtypes)
print("Missing values: ",json_data.isnull().sum().tolist())
print("Duplicate values: ",json_data.duplicated().sum())

No. of rows:  25
No. of Columns:  2
Column Names:  ['country', 'internet_users']
Data types:  country           object
internet_users     int64
dtype: object
Missing values:  [0, 0]
Duplicate values:  0


In [16]:
#Literacy rate


print("No. of rows: ",len(xml_data))
print("No. of Columns: ",len(xml_data.columns))
print("Column Names: ",xml_data.columns.tolist())
print("Data types: ",xml_data.dtypes)
print("Missing values: ",xml_data.isnull().sum().tolist())
print("Duplicate values: ",xml_data.duplicated().sum())

No. of rows:  25
No. of Columns:  2
Column Names:  ['name', 'literacy_rate']
Data types:  name              object
literacy_rate    float64
dtype: object
Missing values:  [0, 0]
Duplicate values:  0


Task 6

Rename columns wherever necessary so that all datasets have a common column name for merging.

Hint:

● Example: Country → country



In [17]:
excel_data.rename(columns = {'Country':'country'}, inplace=True)
xml_data.rename(columns = {'Country':'country'}, inplace=True)

print("CSV (Population):", csv_data.columns.tolist())
print("Excel (GDP):", excel_data.columns.tolist())
print("JSON (Internet Users):", json_data.columns.tolist())
print("XML (Literacy Rate):", xml_data.columns.tolist())

CSV (Population): ['country', 'population']
Excel (GDP): ['country', 'GDP']
JSON (Internet Users): ['country', 'internet_users']
XML (Literacy Rate): ['name', 'literacy_rate']


Task 7

Convert the required columns into appropriate numeric data types.




In [19]:
csv_data['population'] = pd.to_numeric(csv_data['population'])

excel_data['GDP'] = pd.to_numeric(excel_data['GDP'])

json_data['internet_users'] = pd.to_numeric(json_data['internet_users'])

xml_data['literacy_rate'] = pd.to_numeric(xml_data['literacy_rate'])

print("CSV Data Types:\n", csv_data.dtypes, "\n")
print("Excel Data Types:\n", excel_data.dtypes, "\n")
print("JSON Data Types:\n", json_data.dtypes, "\n")
print("XML Data Types:\n", xml_data.dtypes)

CSV Data Types:
 country       object
population     int64
dtype: object 

Excel Data Types:
 country    object
GDP         int64
dtype: object 

JSON Data Types:
 country           object
internet_users     int64
dtype: object 

XML Data Types:
 name              object
literacy_rate    float64
dtype: object


Task 8

Merge all four datasets into one DataFrame.

Hint:

● Use the common column (country)

● Choose an appropriate join method.

In [20]:
xml_data.rename(columns={'name': 'country'}, inplace=True)

merged_df = csv_data.merge(excel_data, on='country', how='inner').merge(json_data, on='country', how='inner').merge(xml_data, on='country', how='inner')

merged_df.head()

,country,population,GDP,internet_users,literacy_rate
0,United States,331002651,30500000,312000000,99.0
1,China,1411778724,19230000,1080000000,97.2
2,India,1380004385,4187000,850000000,77.7
3,Japan,125836021,4180000,118000000,99.0
4,Germany,83240525,4744000,79000000,99.0


Task 9

Create the following calculated columns:

Internet Penetration (%)

GDP Per Capita

Hint:

Round the calculated values to two decimal places.

In [21]:
# 1. Internet Penetration (%) = (internet_users / population) * 100
merged_df['internet_penetration_%'] = ((merged_df['internet_users'] / merged_df['population']) * 100).round(2)

# 2. GDP Per Capita = GDP / population
merged_df['gdp_per_capita'] = (merged_df['GDP'] / merged_df['population']).round(2)

merged_df.head()

,country,population,GDP,internet_users,literacy_rate,internet_penetration_%,gdp_per_capita
0,United States,331002651,30500000,312000000,99.0,94.26,0.09
1,China,1411778724,19230000,1080000000,97.2,76.50,0.01
2,India,1380004385,4187000,850000000,77.7,61.59,0.00
3,Japan,125836021,4180000,118000000,99.0,93.77,0.03
4,Germany,83240525,4744000,79000000,99.0,94.91,0.06


Task 10

Create a SQLite database.

Store the final merged dataset into a table named ‘country_statistics’

Hint:

● Use the sqlite3 library.

● Load the DataFrame using pandas.



In [22]:
import sqlite3

# 1. Connect to SQLite database
conn = sqlite3.connect('country_data.db')

# 2. Store the merged DataFrame into the 'country_statistics' table
merged_df.to_sql('country_statistics', conn, if_exists='replace', index=False)

# 3. Verify data by reading it back from the SQLite database
df_from_db = pd.read_sql_query("SELECT * FROM country_statistics", conn)

conn.close()

Task 11

Verify that the data has been loaded successfully.

Hint:

Retrieve the first five records from the SQLite table.

In [23]:
df_from_db.head()

,country,population,GDP,internet_users,literacy_rate,internet_penetration_%,gdp_per_capita
0,United States,331002651,30500000,312000000,99.0,94.26,0.09
1,China,1411778724,19230000,1080000000,97.2,76.50,0.01
2,India,1380004385,4187000,850000000,77.7,61.59,0.00
3,Japan,125836021,4180000,118000000,99.0,93.77,0.03
4,Germany,83240525,4744000,79000000,99.0,94.91,0.06


Part E – Data Analysis Using Pandas

In [24]:
#  1) List the Top 5 countries by GDP.
top1= merged_df.sort_values(by='GDP', ascending=False).head(5)
top1[['country', 'GDP']]

,country,GDP
0,United States,30500000
1,China,19230000
4,Germany,4744000
2,India,4187000
3,Japan,4180000


In [25]:
#  2) List the Top 5 countries by population.

top2 = merged_df.sort_values(by='population', ascending=False).head(5)
top2[['country', 'population']]

,country,population
1,China,1411778724
2,India,1380004385
0,United States,331002651
15,Indonesia,273523615
10,Brazil,212559417


In [26]:
#  3) Calculate the average literacy rate.

avg = merged_df['literacy_rate'].mean()
print("Average literacy rate:", avg)

Average literacy rate: 97.248


In [27]:
#  4) List all countries having a literacy rate greater than 90%.

coun1 = merged_df[merged_df['literacy_rate'] > 90]
coun1[['country', 'literacy_rate']]

,country,literacy_rate
0,United States,99.0
1,China,97.2
3,Japan,99.0
4,Germany,99.0
5,United Kingdom,99.0
6,France,99.0
7,Italy,99.0
8,Canada,99.0
9,Australia,99.0
10,Brazil,93.2


In [28]:
#  5) List all countries having an internet penetration greater than 70%.

coun2 = merged_df[merged_df['internet_penetration_%'] > 70]
coun2[['country', 'internet_penetration_%']]#  5) List all countries having an internet penetration greater than 70%.



,country,internet_penetration_%
0,United States,94.26
1,China,76.50
3,Japan,93.77
4,Germany,94.91
5,United Kingdom,97.22
6,France,94.98
7,Italy,92.62
8,Canada,94.72
9,Australia,93.43
10,Brazil,85.15


In [29]:
#  6) List the Top 5 countries by GDP per capita.

coun3 = merged_df.sort_values(by='gdp_per_capita', ascending=False).head(5)
coun3[['country', 'gdp_per_capita']]

,country,gdp_per_capita
21,Switzerland,0.11
23,Norway,0.11
24,Singapore,0.09
0,United States,0.09
9,Australia,0.07


In [30]:
#  7) Calculate the total population represented in the dataset.

total_pop = merged_df['population'].sum()
print("Total population represented in the dataset:", total_pop)

Total population represented in the dataset: 4720039725


In [31]:
#  8) Calculate the average internet penetration across all countries.

avg_pen = merged_df['internet_penetration_%'].mean()
print("Average internet penetration across all countries:", avg_pen)

Average internet penetration across all countries: 89.6596


In [32]:
#  9) Identify the country having the highest number of internet users.

coun4 = merged_df[merged_df['internet_users'] == merged_df['internet_users'].max()]
coun4[['country', 'internet_users']]

,country,internet_users
1,China,1080000000


In [33]:
#  10) Find the correlation between Literacy Rate and Internet Penetration.
#Hint: Use the appropriate pandas statistical function.

correlation = merged_df['literacy_rate'].corr(merged_df['internet_penetration_%'])
print("Correlation between Literacy Rate and Internet Penetration:", correlation)

Correlation between Literacy Rate and Internet Penetration: 0.77279756813985


In [34]:
#  11) Display the Top 5 countries by GDP.

coun5 = merged_df.sort_values(by='GDP', ascending=False).head(5)
coun5[['country', 'GDP']]

,country,GDP
0,United States,30500000
1,China,19230000
4,Germany,4744000
2,India,4187000
3,Japan,4180000


In [35]:
#  12) Find the average literacy rate.

avg_lt = merged_df['literacy_rate'].mean()
print("Average literacy rate: ",avg_lt)

Average literacy rate:  97.248


In [36]:
#  13) Display all countries having literacy rate above 90%.

coun6 = merged_df[merged_df['literacy_rate'] > 90]
coun6[['country', 'literacy_rate']]

,country,literacy_rate
0,United States,99.0
1,China,97.2
3,Japan,99.0
4,Germany,99.0
5,United Kingdom,99.0
6,France,99.0
7,Italy,99.0
8,Canada,99.0
9,Australia,99.0
10,Brazil,93.2


In [37]:
#  14) Find the Top 5 countries with the highest internet penetration.

coun7 = merged_df.sort_values(by='internet_penetration_%', ascending=False).head(5)
coun7[['country', 'internet_penetration_%']]

,country,internet_penetration_%
21,Switzerland,98.21
23,Norway,97.76
16,Saudi Arabia,97.66
22,Sweden,97.55
20,Netherlands,97.47


In [38]:
#  15) Calculate the total population.

total_pop = merged_df['population'].sum()
print("Total population: ",total_pop)

Total population:  4720039725


In [39]:
#  16) Find the country with the highest GDP per capita.

coun8 = merged_df[merged_df['gdp_per_capita'] == merged_df['gdp_per_capita'].max()]
coun8[['country', 'gdp_per_capita']]

,country,gdp_per_capita
21,Switzerland,0.11
23,Norway,0.11


In [40]:
#  17)  Export the final merged dataset as a CSV file.

merged_df.to_csv('final_merged_dataset.csv', index=False)


In [41]:
#  18) Sort the dataset in descending order of GDP per Capita and display the Top 10
#countries.

ab = merged_df.sort_values(by='gdp_per_capita', ascending=False).head(10)
ab[['country','gdp_per_capita']]

,country,gdp_per_capita
21,Switzerland,0.11
23,Norway,0.11
24,Singapore,0.09
0,United States,0.09
9,Australia,0.07
20,Netherlands,0.07
4,Germany,0.06
8,Canada,0.06
5,United Kingdom,0.06
22,Sweden,0.06


Data Sources: Extracted data across 4 file formats: CSV (population.csv), Excel (gdp.xlsx), JSON (internet_users.json), and XML (literacy_rate.xml).

Data Cleaning: Verified shapes, schema, missing values, and duplicates; standardized primary join key by renaming Country (Excel) and name (XML) to country.

Transformations: Explicitly converted numerical columns (population, GDP, internet_users, literacy_rate) to numeric data types (int64/float64) and merged all 4 datasets using an inner join on country.

Feature Engineering: Derived two key metrics rounded to 2 decimal places: Internet Penetration (%) and GDP Per Capita.

Database Loading: Connected to a local SQLite database (country_data.db), created table country_statistics using to_sql(), and verified records via SELECT * FROM country_statistics LIMIT 5.

Key Insights: Top GDP Per Capita: Switzerland ( 0.11 ) and Norway ( 0.11 ).

Top Internet Users: China ( 1.08  billion).

Correlation: High positive correlation ( 0.77 ) between literacy rate and internet penetration.

